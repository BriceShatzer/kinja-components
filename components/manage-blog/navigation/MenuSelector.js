/* @flow */

import * as React from 'react';
import styled from 'styled-components';

// ICONS
import ChevronUpIcon from '../../icon19/ChevronUp';
import ChevronDownIcon from '../../icon19/ChevronDown';

import { EnsureDefaultTheme } from '../../theme';
import Toggle from '../../hoc/toggle';

import type { ToggleInjectedProps } from '../../hoc/toggle';


export const Wrapper = styled.div`
	position: relative;
`;

const MenuItemList = styled.ul`
	position: absolute;
	width: 100%;
	background-color: ${props => props.theme.color.white};
	border: 1px solid ${props => props.theme.color.lightgray};
	z-index: 999;
`;

const MenuItem = styled.li`
	padding: 14px;
	text-align: left;
	font-size: 16px;
	color: ${props => props.theme.color.darkgray};
	line-height: 19px;
	cursor: pointer;

	&:hover {
		background-color: ${props => props.theme.color.lightgray};
	}
`;

export const SelectedMenu = styled.div`
	display: flex;
	position: relative;
	bottom: -1px;
	height: 47px;
	padding: 14px;
	padding-left: 0;
	border-bottom: 1px solid ${props => props.theme.color.lightgray};
	cursor: pointer;
`;

const SelectedMenuItem = styled.span`
	width: 100%;
	float: left;
	text-align: left;
	font-size: 16px;
	color: ${props => props.theme.color.gray};
	line-height: 19px;
`;


type Props = {
	activeMenu: string,
	menuList: Array<string>,
	onClick: (value: string) => void
} & ToggleInjectedProps;

type State = {
	activeMenu: string,
	isOpen: boolean
}


class MenuSelector extends React.Component<Props, State> {
	constructor(props: Props) {
		super(props);
		this.state = {
			activeMenu: '',
			isOpen: false
		};
	}

	componentWillReceiveProps(nextProps) {
		nextProps.activeMenu && nextProps.activeMenu !== this.state.activeMenu && this.setState({
			activeMenu: nextProps.activeMenu
		});
	}

	componentWillMount() {
		this.setState({
			activeMenu: this.props.activeMenu,
			isOpen: this.props.isOpen
		});
	}

	onSelectorClicked = () => {
		this.props.toggle();
	}

	onMenuItemSelected = event => {
		const selectedValue = event.currentTarget.innerText;

		this.setState(prevState => {
			if (prevState.activeMenu !== selectedValue) {
				window.ga('send', 'event', 'Manage page click', 'Sub nav click', selectedValue);
				return {
					activeMenu: selectedValue
				};
			}
		}, () => {
			this.props.toggle();
			this.props.onClick(selectedValue);
		});
	}

	render() {
		const { menuList, insideReference, isOpen } = this.props;

		return (
			<EnsureDefaultTheme>
				<Wrapper ref={insideReference}>
					<SelectedMenu onClick={this.onSelectorClicked}>
						<SelectedMenuItem>{this.state.activeMenu}</SelectedMenuItem>
						{isOpen ? <ChevronUpIcon /> : <ChevronDownIcon />}
					</SelectedMenu>
					{isOpen &&
						<MenuItemList>
							{menuList.map(menuItem => <MenuItem key={menuItem} onClick={this.onMenuItemSelected}>{menuItem}</MenuItem>)}
						</MenuItemList>
					}
				</Wrapper>
			</EnsureDefaultTheme>
		);
	}
}


export default Toggle(MenuSelector, { isOutsideClickEnabled: true, isDefaultOpen: false });
