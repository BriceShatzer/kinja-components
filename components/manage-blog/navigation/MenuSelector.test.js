import * as React from 'react';
import { mount } from 'enzyme';

import MenuSelector, { SelectedMenu } from './MenuSelector';

const stubElement = ({
	activeMenu,
	menuList,
	onClick = () => {}
} = {}) => (
	<MenuSelector
		activeMenu={activeMenu}
		menuList={menuList}
		onClick={onClick}
	/>
);

const SAMPLE_ACTIVE_MENU = 'Posts';
const SAMPLE_MENU_LIST = ['Posts', 'Story Types', 'Moderation', 'Members', 'Settings'];


describe('<MenuSelector />', () => {
	let wrapper;
	const handler = jest.fn();

	beforeAll(() => {
		wrapper = mount(stubElement({
			activeMenu: SAMPLE_ACTIVE_MENU,
			menuList: SAMPLE_MENU_LIST,
			onClick: handler
		}));
	});

	it('should render properly', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('should open when SelectedMenu is clicked', () => {
		const selectedMenu = wrapper.find(SelectedMenu);
		selectedMenu.simulate('click');

		expect(wrapper.state().isOpen).toEqual(true);
	});

	// TODO write more TESTS!

});