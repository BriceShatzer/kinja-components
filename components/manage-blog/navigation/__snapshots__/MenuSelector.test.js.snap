// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuSelector /> should render properly 1`] = `
.c3 {
  fill: currentColor;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  bottom: -1px;
  height: 47px;
  padding: 14px;
  padding-left: 0;
  border-bottom: 1px solid #E5E5E5;
  cursor: pointer;
}

.c2 {
  width: 100%;
  float: left;
  text-align: left;
  font-size: 16px;
  color: #7D7D7D;
  line-height: 19px;
}

<Toggle(MenuSelector)
  activeMenu="Posts"
  menuList={
    Array [
      "Posts",
      "Story Types",
      "Moderation",
      "Members",
      "Settings",
    ]
  }
  onClick={[MockFunction]}
>
  <ClickOutside(MenuSelector)
    activeMenu="Posts"
    close={[Function]}
    handleClickOutside={[Function]}
    isOpen={false}
    menuList={
      Array [
        "Posts",
        "Story Types",
        "Moderation",
        "Members",
        "Settings",
      ]
    }
    onClick={[MockFunction]}
    toggle={[Function]}
  >
    <MenuSelector
      activeMenu="Posts"
      close={[Function]}
      insideReference={[Function]}
      isOpen={false}
      menuList={
        Array [
          "Posts",
          "Story Types",
          "Moderation",
          "Members",
          "Settings",
        ]
      }
      onClick={[MockFunction]}
      toggle={[Function]}
    >
      <EnsureDefaultTheme>
        <ThemeProvider
          theme={[Function]}
        >
          <MenuSelector__Wrapper>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "MenuSelector__Wrapper-sc-1ladpb5-0",
                    "isStatic": true,
                    "lastClassName": "c0",
                    "rules": Array [
                      "position:relative;",
                    ],
                  },
                  "displayName": "MenuSelector__Wrapper",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "MenuSelector__Wrapper-sc-1ladpb5-0",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={[Function]}
            >
              <div
                className="c0"
              >
                <MenuSelector__SelectedMenu
                  onClick={[Function]}
                >
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "MenuSelector__SelectedMenu-sc-1ladpb5-3",
                          "isStatic": false,
                          "lastClassName": "c1",
                          "rules": Array [
                            "display:flex;position:relative;bottom:-1px;height:47px;padding:14px;padding-left:0;border-bottom:1px solid ",
                            [Function],
                            ";cursor:pointer;",
                          ],
                        },
                        "displayName": "MenuSelector__SelectedMenu",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "MenuSelector__SelectedMenu-sc-1ladpb5-3",
                        "target": "div",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onClick={[Function]}
                  >
                    <div
                      className="c1"
                      onClick={[Function]}
                    >
                      <MenuSelector__SelectedMenuItem>
                        <StyledComponent
                          forwardedComponent={
                            Object {
                              "$$typeof": Symbol(react.forward_ref),
                              "attrs": Array [],
                              "componentStyle": ComponentStyle {
                                "componentId": "MenuSelector__SelectedMenuItem-sc-1ladpb5-4",
                                "isStatic": false,
                                "lastClassName": "c2",
                                "rules": Array [
                                  "width:100%;float:left;text-align:left;font-size:16px;color:",
                                  [Function],
                                  ";line-height:19px;",
                                ],
                              },
                              "displayName": "MenuSelector__SelectedMenuItem",
                              "foldedComponentIds": Array [],
                              "render": [Function],
                              "styledComponentId": "MenuSelector__SelectedMenuItem-sc-1ladpb5-4",
                              "target": "span",
                              "toString": [Function],
                              "warnTooManyClasses": [Function],
                              "withComponent": [Function],
                            }
                          }
                          forwardedRef={null}
                        >
                          <span
                            className="c2"
                          >
                            Posts
                          </span>
                        </StyledComponent>
                      </MenuSelector__SelectedMenuItem>
                      <ChevronDownIcon>
                        <Icon19>
                          <icon19__IconWrapper>
                            <StyledComponent
                              forwardedComponent={
                                Object {
                                  "$$typeof": Symbol(react.forward_ref),
                                  "attrs": Array [],
                                  "componentStyle": ComponentStyle {
                                    "componentId": "icon19__IconWrapper-iyvn34-0",
                                    "isStatic": true,
                                    "lastClassName": "c3",
                                    "rules": Array [
                                      "fill:currentColor;display:flex;",
                                    ],
                                  },
                                  "displayName": "icon19__IconWrapper",
                                  "foldedComponentIds": Array [],
                                  "render": [Function],
                                  "styledComponentId": "icon19__IconWrapper-iyvn34-0",
                                  "target": "span",
                                  "toString": [Function],
                                  "warnTooManyClasses": [Function],
                                  "withComponent": [Function],
                                }
                              }
                              forwardedRef={null}
                            >
                              <span
                                className="c3"
                              >
                                <ChevronDown
                                  aria-label="ChevronDown icon"
                                  height="18"
                                  viewBox="0 0 18 18"
                                  width="18"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <svg
                                    aria-label="ChevronDown icon"
                                    height="18"
                                    viewBox="0 0 18 18"
                                    width="18"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M3.87 6.16a.5.5 0 0 0-.74.68l5.5 6c.2.21.54.21.74 0l5.5-6a.5.5 0 1 0-.74-.68L9 11.76l-5.13-5.6z"
                                      fillRule="evenodd"
                                    />
                                  </svg>
                                </ChevronDown>
                              </span>
                            </StyledComponent>
                          </icon19__IconWrapper>
                        </Icon19>
                      </ChevronDownIcon>
                    </div>
                  </StyledComponent>
                </MenuSelector__SelectedMenu>
              </div>
            </StyledComponent>
          </MenuSelector__Wrapper>
        </ThemeProvider>
      </EnsureDefaultTheme>
    </MenuSelector>
  </ClickOutside(MenuSelector)>
</Toggle(MenuSelector)>
`;
